File: src/LumeHub.Server/SmartHome/Request.cs
namespace LumeHub.Server.SmartHome;

public sealed class Request
{
    public required string RequestId { get; init; }
    public required List<SmartHomeInput>? Inputs { get; init; }
}

public sealed class SmartHomeInput
{
    public required string Intent { get; init; }
    public SmartHomePayload? Payload { get; init; }
}

public sealed class SmartHomePayload
{
    public List<SmartHomeCommand>? Commands { get; init; }
}

public sealed class SmartHomeCommand
{
    public List<Dictionary<string, string>>? Devices { get; init; }
    public List<SmartHomeExecution>? Execution { get; init; }
}

public sealed class SmartHomeExecution
{
    public required string Command { get; init; }
    public SmartHomeExecutionParams? Params { get; init; }
}

public sealed class SmartHomeExecutionParams
{
    public bool On { get; init; }
}

File: crates/api/google/tests/request.rs
use api_google::request::{ExecuteCommandType, Intent, SmartHomeRequest};

#[test]
fn intent_sync_deserializes() {
    let json = r#"{"requestId":"r1","inputs":[{"intent":"action.devices.SYNC"}]}"#;
    let req: SmartHomeRequest = serde_json::from_str(json).unwrap();
    assert_eq!(req.inputs[0].intent, Intent::Sync);
}

#[test]
fn intent_execute_deserializes() {
    let json = r#"{"requestId":"r2","inputs":[{"intent":"action.devices.EXECUTE","payload":{"commands":[]}}]}"#;
    let req: SmartHomeRequest = serde_json::from_str(json).unwrap();
    assert_eq!(req.inputs[0].intent, Intent::Execute);
}

#[test]
fn execute_command_type_on_off_deserializes() {
    let json = r#""action.devices.commands.OnOff""#;
    let cmd: ExecuteCommandType = serde_json::from_str(json).unwrap();
    assert_eq!(cmd, ExecuteCommandType::OnOff);
}

#[test]
fn command_request_get_params_extracts_value() {
    use api_google::request::{CommandRequest, DeviceRequest, ExecutionRequest};
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct OnOff {
        on: bool,
    }

    let cmd = CommandRequest {
        devices: vec![DeviceRequest {
            id: "led-strip".to_string(),
        }],
        execution: vec![ExecutionRequest {
            command: ExecuteCommandType::OnOff,
            params: serde_json::json!({"on": true}),
        }],
    };

    let params: OnOff = cmd.get_params().unwrap();
    assert!(params.on);
}

#[test]
fn command_request_get_params_errors_on_empty_execution() {
    use api_google::request::{CommandRequest, DeviceRequest};
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Dummy;

    let cmd = CommandRequest {
        devices: vec![DeviceRequest {
            id: "x".to_string(),
        }],
        execution: vec![],
    };

    assert!(cmd.get_params::<Dummy>().is_err());
}
